core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -621,6 +544,114 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                parser.getTokenLocation().lineNumber, parser.getTokenLocation().columnNumber);
+        }
+
+        builder.put(innerBuilder.build());
+        return builder;
+    }
+
+    private static void fromXContent(XContentParser parser, StringBuilder keyBuilder, Settings.Builder builder,
+                                     boolean allowNullValues) throws IOException {
+        final int length = keyBuilder.length();
+        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            if (parser.currentToken() == XContentParser.Token.FIELD_NAME) {
+                keyBuilder.setLength(length);
+                keyBuilder.append(parser.currentName());
+            } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                keyBuilder.append('.');
+                fromXContent(parser, keyBuilder, builder, allowNullValues);
+            } else if (parser.currentToken() == XContentParser.Token.START_ARRAY) {
+                List<String> list = new ArrayList<>();
+                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                    if (parser.currentToken() == XContentParser.Token.VALUE_STRING) {
+                        list.add(parser.text());
+                    } else if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+                        list.add(parser.text()); // just use the string representation here
+                    } else if (parser.currentToken() == XContentParser.Token.VALUE_BOOLEAN) {
+                        list.add(String.valueOf(parser.text()));
+                    } else {
+                        throw new IllegalStateException("only value lists are allowed in serialized settings");
+                    }
+                }
+                String key = keyBuilder.toString();
+                valdiateValue(key, list, builder, parser, allowNullValues);
+                builder.putArray(key, list);
+            } else if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                String key = keyBuilder.toString();
+                valdiateValue(key, null, builder, parser, allowNullValues);
+                builder.putNull(key);
+            } else if (parser.currentToken() == XContentParser.Token.VALUE_STRING
+                || parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+                String key = keyBuilder.toString();
+                String value = parser.text();
+                valdiateValue(key, value, builder, parser, allowNullValues);
+                builder.put(key, value);
+            } else if (parser.currentToken() == XContentParser.Token.VALUE_BOOLEAN) {
+                String key = keyBuilder.toString();
+                valdiateValue(key, parser.text(), builder, parser, allowNullValues);
+                builder.put(key, parser.booleanValue());
+            } else {
+                throw new IllegalStateException("Illegal token: " + parser.currentToken());
```

Comment:
You can use `XContentParserUtils.throwUnknownToken()` (that would throw a ParsingException instead but I think it's appropriate here)